### PR TITLE
New feature: Global phase linear constraints 

### PR DESCRIPTION
**Pull request by szabino-mrszabo**

### CODE REVIEW

File: src/data.jl [hunk 1]
```diff
@@ -255,10 +259,17 @@ function get_target_gate(params::Dict{String, Any}, are_elementary_gates_real::B
...
-            Memento.error(_LOGGER, "Infeasible decomposition: all elementary gates have zero imaginary parts")
-        else 
+        if is_target_real
             return real(params["target_gate"]), is_target_real
+        else
+            ref_nonzero_r, ref_nonzero_c = QCO._get_ref_nonzero_index_of_original_target(params["target_gate"])
+            global_phase = angle(params["target_gate"][ref_nonzero_r, ref_nonzero_c])
+            is_target_real_up_to_phase = QCO.is_gate_real(exp(-im*global_phase)*params["target_gate"])
+            if is_target_real_up_to_phase
+                return real(exp(-im*global_phase)*params["target_gate"]), !is_target_real
```
Comment (harshangrjn) [inline src/data.jl:269]: Could this also be written as a general evaluation for `global_phase = 0` and then evaluate the global phase of target if `exact_optimal_global_phase` is true? Also, we should add a test to cover this case where target is real in a `global_phase` sense.

File: src/log.jl [hunk 2]
```diff
@@ -193,10 +193,26 @@ function validate_circuit_decomposition(data::Dict{String, Any}, id_sequence::Ar
         target_gate = QCO.real_to_complex_gate(data["target_gate"])
     end
     
-    if (data["decomposition_type"] in ["exact_optimal", "exact_feasible"]) && (!isapprox(M_sol, target_gate, atol = 1E-4))
-        Memento.error(_LOGGER, "Decomposition is not valid: Problem may be infeasible")
+    global_phase = 1
+
+    if data["decomposition_type"] in ["exact_optimal_global_phase"]
+        n_r = size(target_gate)[1]
```
Comment (harshangrjn) [inline src/log.jl:199]: Could these refs be directly obtained using your helper function in utility.jl?

File: src/constraints.jl [hunk 3]
```diff
@@ -150,12 +150,64 @@ function constraint_gate_target_condition_compact(qcm::QuantumCircuitModel)
...
+    ref_nonzero_r, ref_nonzero_c = QCO._get_ref_nonzero_index_of_real_target(qcm.data::Dict{String,Any})
+
+    for i=1:n_r, j=1:n_c
+        if isapprox(qcm.data["target_gate"][i,j], 0, atol=1E-6)
+            JuMP.@constraint(qcm.model, U_var[i,j,max_depth] == 0)
+        else
+            JuMP.@constraint(qcm.model, U_var[i,j,max_depth]*qcm.data["target_gate"][ref_nonzero_r,ref_nonzero_c] == qcm.data["target_gate"][i,j]*U_var[ref_nonzero_r,ref_nonzero_c,max_depth])
+        end
+    end
+    return
```
Comment (harshangrjn) [inline src/constraints.jl:182]: This function needs to be tested in unit tests. Also, I think this and `constraint_gate_target_condition_compact_global_phase_complex` could be combined into a single function?

**Review by harshangrjn — COMMENTED** [review]
Added comments

File: docs/src/quickguide.md [hunk 4]
```diff
@@ -46,7 +46,7 @@ QCOpt takes two types of user-defined input specifications. The first type of in
...
-| `decomposition_type` | Choose one of the following: (a) `exact_optimal`, which finds an exact, provably optimal, decomposition if it exists, (c) `exact_feasible`, which finds any feasible exact decomposition, but not necessarily an optimal one if it exsists (c) `approximate`, which finds an approximate decomposition if an exact one does not exist; otherwise it will return an exact decomposition (default: `exact_optimal`)|
+| `decomposition_type` | Choose one of the following: (a) `exact_optimal`, which finds an exact, provably optimal, decomposition if it exists, (b) `exact_feasible`, which finds any feasible exact decomposition, but not necessarily an optimal one if it exsist, (c) `exact_optimal_global_phase`, which finds an optimal circuit decomposition if it exists, up to a global phase (d) `approximate`, which finds an approximate decomposition if an exact one does not exist; otherwise it will return an exact decomposition (default: `exact_optimal`)|
```
[the remark this answers is not in the document]
Comment (szabino-mrszabo) [reply docs/src/quickguide.md:49]: including this in the new commit

File: src/data.jl (hunk 1 above)
Comment (szabino-mrszabo) [reply src/data.jl:269]: implemented the global_phase 0, added tests  and including this in the new commit

File: src/constraints.jl (hunk 3 above)
Comment (szabino-mrszabo) [reply src/constraints.jl:182]: combined the two global_phase_constraints
unit test now covering both cases

File: src/utility.jl [hunk 5]
```diff
@@ -717,13 +717,60 @@ function _determinant_test_for_infeasibility(data::Dict{String,Any})
...
 
+function _get_ref_nonzero_index_of_real_target(data::Dict{String,Any})
+
+    n_r            = size(data["gates_real"])[1]
+    n_c            = size(data["gates_real"])[2]
+
+    ref_nonzero_r = 0
+    ref_nonzero_c = 0
+    if data["are_gates_real"]
+        for i=1:n_r, j=1:n_c
```
[the remark this answers is not in the document]
Comment (szabino-mrszabo) [reply src/utility.jl:728]: added unit test to cover all-real case as well

File: src/log.jl (hunk 2 above)
Comment (szabino-mrszabo) [reply src/log.jl:199]: changed so this uses: _get_ref_nonzero_index_of_original_target

File: src/utility.jl [hunk 6]
```diff
@@ -717,13 +717,60 @@ function _determinant_test_for_infeasibility(data::Dict{String,Any})
...
+                ref_nonzero_c = j
+                break
+            end
+        end
+    end
+
+    return ref_nonzero_r, ref_nonzero_c 
+end
+
+function _get_ref_nonzero_index_of_original_target(T::Array{Complex{Float64},2})
```
[the remark this answers is not in the document]
Comment (szabino-mrszabo) [reply src/utility.jl:748]: I added a unit test for this as well. 

I see where you are coming from; I could add an option to make it one function but I feel it is probably easier to understand to have 2 separate functions for 2 things that I consider very different objects. Like apples and pears.
+ One is just a complex array 'target' as it is an input
+ One is a QuantumCircuitModel.data.target_gate thing which is under the hood a real array but nonetheless is this part of this magical type QuantumCircuitModel

let me know what you think.

**Comment by harshangrjn** [comment]
Thanks @szabino-mrszabo

**Pull request merged**